# SETUP AND IMPORT LIBRARY

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [15]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
import seaborn as sns
import matplotlib.pyplot as plt
import math
from scipy.stats import skew
from sklearn.preprocessing import PowerTransformer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [ ]:
SEED = 42

# LOAD DATASET

In [ ]:
train_set = pd.read_csv("/content/drive/MyDrive/Datathon 2025 (HidupJ0kow1)/Dataset/UNSW_NB15_training-set.csv")
test_set = pd.read_csv("/content/drive/MyDrive/Datathon 2025 (HidupJ0kow1)/Dataset/UNSW_NB15_testing-set.csv")
event_list = pd.read_csv("/content/drive/MyDrive/Datathon 2025 (HidupJ0kow1)/Dataset/UNSW-NB15_LIST_EVENTS.csv")
nb15_1 = pd.read_csv("/content/drive/MyDrive/Datathon 2025 (HidupJ0kow1)/Dataset/UNSW-NB15_1.csv")
nb15_2 = pd.read_csv("/content/drive/MyDrive/Datathon 2025 (HidupJ0kow1)/Dataset/UNSW-NB15_2.csv")
nb15_3 = pd.read_csv("/content/drive/MyDrive/Datathon 2025 (HidupJ0kow1)/Dataset/UNSW-NB15_3.csv")
nb15_4 = pd.read_csv("/content/drive/MyDrive/Datathon 2025 (HidupJ0kow1)/Dataset/UNSW-NB15_4.csv")
features = pd.read_csv("/content/drive/MyDrive/Datathon 2025 (HidupJ0kow1)/Dataset/NUSW-NB15_features.csv", encoding='latin1')


In [ ]:
event_list

Attack category Attack subcategory  Number of events
0            normal                NaN           2218761
1          Fuzzers                 FTP               558
2          Fuzzers                HTTP              1497
3          Fuzzers                 RIP              3550
4          Fuzzers                 SMB              5245
..              ...                ...               ...
203       Shellcode           Decoders                18
204       Shellcode        Multiple OS                10
205       Shellcode            Solaris                12
206           Worms                                  174
207             NaN             Total            2540044

[208 rows x 3 columns]

In [ ]:
nb15_1.head()

59.166.0.0   1390  149.171.126.6    53  udp  CON  0.001055  132  164  31  \
0  59.166.0.0  33661  149.171.126.9  1024  udp  CON  0.036133  528  304  31   
1  59.166.0.6   1464  149.171.126.7    53  udp  CON  0.001119  146  178  31   
2  59.166.0.5   3593  149.171.126.5    53  udp  CON  0.001209  132  164  31   
3  59.166.0.3  49664  149.171.126.0    53  udp  CON  0.001169  146  178  31   
4  59.166.0.0  32119  149.171.126.9   111  udp  CON  0.078339  568  312  31   

   ...  0.17   3  7  1  3.1  1.1  1.2  1.3  Unnamed: 47  0.18  
0  ...     0   2  4  2    3    1    1    2          NaN     0  
1  ...     0  12  8  1    2    2    1    1          NaN     0  
2  ...     0   6  9  1    1    1    1    1          NaN     0  
3  ...     0   7  9  1    1    1    1    1          NaN     0  
4  ...     0   2  4  2    3    1    1    2          NaN     0  

[5 rows x 49 columns]

In [ ]:
nb15_2.head()

59.166.0.0   6055  149.171.126.5  54145  tcp  FIN  0.072974  4238  60788  \
0  59.166.0.0   7832  149.171.126.3   5607  tcp  FIN  0.144951  5174  91072   
1  59.166.0.8  11397  149.171.126.6     21  tcp  FIN  0.116107  2934   3742   
2  59.166.0.0   3804  149.171.126.3     53  udp  CON  0.000986   146    178   
3  59.166.0.8  14339  149.171.126.6  14724  tcp  FIN  0.038480  8928    320   
4  59.166.0.8  39094  149.171.126.3     53  udp  CON  0.001026   130    162   

   31  ...  0.6  13  13.1  6  7.1  1  1.1  2  Unnamed: 47  0.7  
0  31  ...    0  13    13  6    7  1    1  2          NaN    0  
1  31  ...    1   1     2  7    5  1    1  4          NaN    0  
2  31  ...    0  13    13  6    7  1    1  2          NaN    0  
3  31  ...    0   8    20  7    5  1    1  4          NaN    0  
4  31  ...    0   8    13  6    5  1    1  1          NaN    0  

[5 rows x 49 columns]

In [ ]:
nb15_3.head()

59.166.0.1  18247  149.171.126.4   7662  tcp  FIN  0.119596   4550  \
0      59.166.0.3  54771  149.171.126.2  27709  tcp  FIN  0.650574   8928   
1      59.166.0.8  13289  149.171.126.9   5190  tcp  FIN  0.007980   2158   
2  149.171.126.18   1043   175.45.176.3     53  udp  INT  0.000005    264   
3  149.171.126.18   1043   175.45.176.3     53  udp  INT  0.000005    264   
4      59.166.0.3  10275  149.171.126.0     25  tcp  FIN  0.486578  37462   

   68342  31  ...      6   2 2.1   5   1  1.1  2.2  Unnamed: 47  0.4  
0    320  31  ...      3   5   2   4   1    1    4          NaN    0  
1   2464  31  ...      3   5   1   1   1    1    3          NaN    0  
2      0  60  ...     19  19  19  19  19   19   19          NaN    0  
3      0  60  ...     19  19  19  19  19   19   19          NaN    0  
4   3380  31  ...      3   2   3   4   2    1    2          NaN    0  

[5 rows x 49 columns]

In [ ]:
nb15_4.head()

59.166.0.9   7045  149.171.126.7    25  tcp  FIN  0.201886  37552     3380  \
0  59.166.0.9   9685  149.171.126.2    80  tcp  FIN  5.864748  19410  1087890   
1  59.166.0.2   1421  149.171.126.4    53  udp  CON  0.001391    146      178   
2  59.166.0.2  21553  149.171.126.2    25  tcp  FIN  0.053948  37812     3380   
3  59.166.0.8  45212  149.171.126.4    53  udp  CON  0.000953    146      178   
4  59.166.0.0  59922  149.171.126.8  6881  tcp  FIN  8.633186  25056  1094788   

   31  ...     2  2.1  7  4  1  1.1  3  Unnamed: 47  0.4  
0  31  ...     3    1  4  4  1    1  1          NaN    0  
1  31  ...     3    5  2  7  1    1  4          NaN    0  
2  31  ...     1    1  4  7  1    1  3          NaN    0  
3  31  ...     2    5  2  1  1    1  2          NaN    0  
4  31  ...     9    7  2  3  2    1  6          NaN    0  

[5 rows x 49 columns]

In [ ]:
features

No.              Name      Type   \
0     1             srcip    nominal   
1     2             sport    integer   
2     3             dstip    nominal   
3     4            dsport    integer   
4     5             proto    nominal   
5     6             state    nominal   
6     7               dur      Float   
7     8            sbytes    Integer   
8     9            dbytes    Integer   
9    10              sttl    Integer   
10   11              dttl    Integer   
11   12             sloss    Integer   
12   13             dloss    Integer   
13   14           service    nominal   
14   15             Sload      Float   
15   16             Dload      Float   
16   17             Spkts    integer   
17   18             Dpkts    integer   
18   19              swin    integer   
19   20              dwin    integer   
20   21             stcpb    integer   
21   22             dtcpb    integer   
22   23           smeansz    integer   
23   24           dmeansz    integer   
24   25       trans_depth    integer   
25   26       res_bdy_len    integer   
26   27              Sjit      Float   
27   28              Djit      Float   
28   29             Stime  Timestamp   
29   30             Ltime  Timestamp   
30   31           Sintpkt      Float   
31   32           Dintpkt      Float   
32   33            tcprtt      Float   
33   34            synack      Float   
34   35            ackdat      Float   
35   36   is_sm_ips_ports     Binary   
36   37      ct_state_ttl    Integer   
37   38  ct_flw_http_mthd    Integer   
38   39      is_ftp_login     Binary   
39   40        ct_ftp_cmd    integer   
40   41        ct_srv_src    integer   
41   42        ct_srv_dst    integer   
42   43        ct_dst_ltm    integer   
43   44       ct_src_ ltm    integer   
44   45  ct_src_dport_ltm    integer   
45   46  ct_dst_sport_ltm    integer   
46   47    ct_dst_src_ltm    integer   
47   48        attack_cat    nominal   
48   49             Label     binary   

                                          Description  
0                                   Source IP address  
1                                  Source port number  
2                              Destination IP address  
3                             Destination port number  
4                                Transaction protocol  
5   Indicates to the state and its dependent proto...  
6                               Record total duration  
7            Source to destination transaction bytes   
8             Destination to source transaction bytes  
9           Source to destination time to live value   
10           Destination to source time to live value  
11           Source packets retransmitted or dropped   
12       Destination packets retransmitted or dropped  
13  http, ftp, smtp, ssh, dns, ftp-data ,irc  and ...  
14                             Source bits per second  
15                        Destination bits per second  
16                Source to destination packet count   
17                 Destination to source packet count  
18              Source TCP window advertisement value  
19         Destination TCP window advertisement value  
20                    Source TCP base sequence number  
21               Destination TCP base sequence number  
22  Mean of the ?ow packet size transmitted by the...  
23  Mean of the ?ow packet size transmitted by the...  
24  Represents the pipelined depth into the connec...  
25  Actual uncompressed content size of the data t...  
26                               Source jitter (mSec)  
27                          Destination jitter (mSec)  
28                                  record start time  
29                                   record last time  
30             Source interpacket arrival time (mSec)  
31        Destination interpacket arrival time (mSec)  
32  TCP connection setup round-trip time, the sum ...  
33  TCP connection setup time, the time between th...  
34  TCP connection setup time, the time bet

### CONCATE THE DATA

In [ ]:
for df in (nb15_1, nb15_2, nb15_3, nb15_4):
    df.columns = features['Name']

In [ ]:
train_df = pd.concat([nb15_1, nb15_2, nb15_3, nb15_4],
                     ignore_index=True)

# BASIC DATA PREPROCESSING

In [ ]:
# SHUFFLING THE DATA
train_df = train_df.sample(frac=1, random_state=SEED).reset_index(drop=True)

In [ ]:
# DROP DUPLICATES
train_df = train_df.drop_duplicates().reset_index(drop=True)

In [ ]:
train_df.isnull().sum()


Name
srcip                     0
sport                     0
dstip                     0
dsport                    0
proto                     0
state                     0
dur                       0
sbytes                    0
dbytes                    0
sttl                      0
dttl                      0
sloss                     0
dloss                     0
service                   0
Sload                     0
Dload                     0
Spkts                     0
Dpkts                     0
swin                      0
dwin                      0
stcpb                     0
dtcpb                     0
smeansz                   0
dmeansz                   0
trans_depth               0
res_bdy_len               0
Sjit                      0
Djit                      0
Stime                     0
Ltime                     0
Sintpkt                   0
Dintpkt                   0
tcprtt                    0
synack                    0
ackdat                    0
is_sm_ips_ports           0
ct_state_ttl              0
ct_flw_http_mthd     933603
is_ftp_login        1014165
ct_ftp_cmd                0
ct_srv_src                0
ct_srv_dst                0
ct_dst_ltm                0
ct_src_ ltm               0
ct_src_dport_ltm          0
ct_dst_sport_ltm          0
ct_dst_src_ltm            0
attack_cat          1959771
Label                     0
dtype: int64

In [ ]:
missing_info = pd.DataFrame({
    'Missing Values': train_df.isnull().sum(),
    'Percentage (%)': train_df.isnull().mean() * 100
}).sort_values(by='Missing Values', ascending=False)

missing_info.head()  # Tampilkan 5 kolom dengan missing terbanyak


Missing Values  Percentage (%)
Name                                            
attack_cat               1959771       95.161446
is_ftp_login             1014165       49.245248
ct_flw_http_mthd          933603       45.333364
sport                          0        0.000000
proto                          0        0.000000

### Handle Missing Values

Dalam proses pembersihan data, beberapa nilai hilang (NaN) ditangani berdasarkan makna kontekstualnya. Pada kolom attack_cat, nilai NaN diartikan sebagai data yang tidak termasuk dalam kategori serangan, sehingga diisi dengan nilai 'normal' untuk menjaga konsistensi informasi serta mempermudah klasifikasi. Selanjutnya, pada kolom ct_flw_http_mthd, nilai NaN menunjukkan tidak adanya aliran (flow) dengan metode HTTP tertentu dalam sesi tersebut, sehingga digantikan dengan nilai 0. Hal serupa juga berlaku untuk kolom is_ftp_login, di mana ketiadaan data mengindikasikan bahwa tidak terjadi proses login FTP, sehingga nilai NaN juga diisi dengan 0. Pendekatan ini dilakukan agar data tetap utuh tanpa perlu menghapus baris, sekaligus memastikan bahwa model tidak salah mengartikan nilai kosong sebagai noise atau outlier.

In [ ]:
train_df['attack_cat'].unique()

array([nan, 'Generic', 'Exploits', ' Reconnaissance ', ' Fuzzers ', 'DoS',
       'Analysis', 'Backdoor', 'Reconnaissance', ' Fuzzers', 'Backdoors',
       'Worms', ' Shellcode ', 'Shellcode'], dtype=object)

In [ ]:
# PERBAIKI attack_cat
train_df['attack_cat'] = (
    train_df['attack_cat']
    .str.strip()
    .str.lower()
)

In [ ]:
# NaN → dianggap bukan serangan, jadi diisi 'normal'
train_df['attack_cat'] = train_df['attack_cat'].fillna('normal') \
                                               .str.strip() \
                                               .str.lower()

In [ ]:
# MAPPING LABEL SESUAI DI CSV FEATURES
mapping = {
    'generic':        'Generic',
    'exploit':        'DoS Exploits',
    'exploits':       'DoS Exploits',
    'dos':            'DoS Exploits',
    'fuzzer':         'Fuzzers',
    'fuzzers':        'Fuzzers',
    'analysis':       'Analysis',
    'backdoor':       'Backdoors',
    'backdoors':      'Backdoors',
    'reconnaissance': 'Reconnaissance',
    'shellcode':      'Shellcode',
    'worm':           'Worms',
    'worms':          'Worms',
    'normal':         'Normal'
}
train_df['attack_cat'] = train_df['attack_cat'].map(mapping)

In [ ]:
print("Unique attack_cat:", train_df['attack_cat'].unique())
print("Missing attack_cat after map:", train_df['attack_cat'].isna().sum())

Unique attack_cat: ['Normal' 'Generic' 'DoS Exploits' 'Reconnaissance' 'Fuzzers' 'Analysis'
 'Backdoors' 'Worms' 'Shellcode']
Missing attack_cat after map: 0


In [ ]:
train_df['ct_flw_http_mthd'] = train_df['ct_flw_http_mthd'].fillna(0).astype(int)
train_df['is_ftp_login']     = train_df['is_ftp_login'].fillna(0).astype(int)
train_df['ct_ftp_cmd']       = (
    pd.to_numeric(train_df['ct_ftp_cmd'], errors='coerce')
      .fillna(0)
      .astype(int)
)
print("Missing counts:", train_df[['ct_flw_http_mthd','is_ftp_login','ct_ftp_cmd']].isna().sum().to_dict())

Missing counts: {'ct_flw_http_mthd': 0, 'is_ftp_login': 0, 'ct_ftp_cmd': 0}


In [ ]:
# STANDARISASI HEADER
train_df.columns = (
    train_df.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace('-', '_')
)

In [ ]:
#  STRIP & LOWER object cols (tanpa fillna global)
for c in train_df.select_dtypes(include='object').columns:
    train_df[c] = train_df[c].astype(str).str.strip().str.lower()

## HANDLING WRONG DATA TYPE

In [ ]:
# Kolom nominal tetap object/string
nominal_cols = ['srcip', 'dstip', 'proto', 'state', 'service', 'attack_cat']
for c in nominal_cols:
    if c in train_df:
        train_df[c] = train_df[c].astype(str)

In [ ]:
# Kolom integer
int_cols = [
    'sport','dsport','sttl','dttl','sloss','dloss','spkts','dpkts',
    'swin','dwin','stcpb','dtcpb','smeansz','dmeansz','trans_depth',
    'res_bdy_len','is_sm_ips_ports','ct_state_ttl','ct_flw_http_mthd',
    'is_ftp_login','ct_ftp_cmd','ct_srv_src','ct_srv_dst','ct_dst_ltm',
    'ct_src__ltm','ct_src_dport_ltm','ct_dst_sport_ltm','ct_dst_src_ltm',
    'label', 'sbytes','dbytes'
]
for c in int_cols:
    if c in train_df:
        train_df[c] = pd.to_numeric(train_df[c], errors='coerce').fillna(0).astype(int)

In [ ]:
# Kolom float
float_cols = [
    'dur', 'sload','dload','sjit','djit',
    'sintpkt','dintpkt','tcprtt','synack','ackdat'
]
for c in float_cols:
    if c in train_df:
        train_df[c] = pd.to_numeric(train_df[c], errors='coerce').astype(float)

In [ ]:
# Timestamp
time_cols = ['stime','ltime']
for c in time_cols:
    if c in train_df:
        train_df[c] = pd.to_datetime(train_df[c], unit='s', errors='coerce')

In [ ]:
# Cek dulu nilai unik
print("Before cleaning:", train_df['is_ftp_login'].unique())

# Hapus baris dengan nilai tidak valid (jika hanya 0 dan 1 yang boleh)
train_df = train_df[train_df['is_ftp_login'].isin([0, 1])]

# Pastikan tipenya integer
train_df['is_ftp_login'] = train_df['is_ftp_login'].astype(int)

# Cek hasil akhirnya
print("After cleaning:", train_df['is_ftp_login'].unique())


Before cleaning: [0 1 4 2]
After cleaning: [0 1]


In [ ]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2059312 entries, 0 to 2059416
Data columns (total 49 columns):
 #   Column            Dtype         
---  ------            -----         
 0   srcip             object        
 1   sport             int64         
 2   dstip             object        
 3   dsport            int64         
 4   proto             object        
 5   state             object        
 6   dur               float64       
 7   sbytes            int64         
 8   dbytes            int64         
 9   sttl              int64         
 10  dttl              int64         
 11  sloss             int64         
 12  dloss             int64         
 13  service           object        
 14  sload             float64       
 15  dload             float64       
 16  spkts             int64         
 17  dpkts             int64         
 18  swin              int64         
 19  dwin              int64         
 20  stcpb             int64         
 21  dtcpb        

In [ ]:
# (1) Banyak duplikat sebelum dihapus
n_dupes = train_df.duplicated().sum()
print(f"Ada {n_dupes} baris duplikat sebelum drop.")

# (2) Hapus semua duplikat dan reset index
train_df = train_df.drop_duplicates().reset_index(drop=True)

# (3) Cek lagi duplikat setelah dihapus
print("Setelah drop:", train_df.duplicated().sum(), "duplikat")

Ada 2000 baris duplikat sebelum drop.
Setelah drop: 0 duplikat


In [ ]:
label_counts = train_df['label'].value_counts()
label_percentages = train_df['label'].value_counts(normalize=True) * 100

print("Jumlah tiap label:")
print(label_counts)
print("\nPersentase tiap label:")
print(label_percentages.round(2))


Jumlah tiap label:
label
0    1959668
1      97644
Name: count, dtype: int64

Persentase tiap label:
label
0    95.25
1     4.75
Name: proportion, dtype: float64


In [ ]:
def generate_features(df):
    df = df.copy()
    eps = 1e-5  # untuk hindari div/0

    # — 1) Pastikan stime & ltime datetime
    df['stime'] = pd.to_datetime(df['stime'], errors='coerce')
    df['ltime'] = pd.to_datetime(df['ltime'], errors='coerce')

    # — 2) Duration (detik), clip ke >=0
    df['duration'] = (
        df['ltime'] - df['stime']
    ).dt.total_seconds().clip(lower=0)

    # — 3) Rate features (per detik)
    df['src_bps'] = df['sbytes'] / (df['duration'] + eps)
    df['dst_bps'] = df['dbytes'] / (df['duration'] + eps)
    df['src_pps'] = df['spkts'] / (df['duration'] + eps)
    df['dst_pps'] = df['dpkts'] / (df['duration'] + eps)

    # — 4) Ratio features
    df['byte_ratio']       = df['sbytes'] / (df['dbytes'] + eps)
    df['pkt_ratio']        = df['spkts'] / (df['dpkts'] + eps)
    df['load_ratio']       = df['sload']  / (df['dload'] + eps)
    df['jit_ratio']        = df['sjit']   / (df['djit']  + eps)
    df['inter_pkt_ratio']  = df['sintpkt']/ (df['dintpkt']+ eps)
    df['tcp_setup_ratio']  = df['tcprtt'] / (df['synack'] + df['ackdat'] + eps)

    # — 5) Outlier‐robust loss/window ratios
    df['loss_ratio_src']   = df['sloss']  / (df['spkts'] + eps)
    df['loss_ratio_dst']   = df['dloss']  / (df['dpkts'] + eps)
    df['window_ratio']     = df['swin']   / (df['dwin']  + eps)

    # — 6) Aggregates
    df['total_bytes']      = df['sbytes'] + df['dbytes']
    df['total_pkts']       = df['spkts']  + df['dpkts']
    df['total_load']       = df['sload']  + df['dload']
    df['total_jitter']     = df['sjit']   + df['djit']
    df['total_inter_pkt']  = df['sintpkt']+ df['dintpkt']
    df['total_tcp_setup']  = df['tcprtt'] + df['synack'] + df['ackdat']

    # — 7) Interaction features
    df['byte_pkt_interaction_src'] = df['sbytes'] * df['spkts']
    df['byte_pkt_interaction_dst'] = df['dbytes'] * df['dpkts']
    df['load_jit_interaction_src'] = df['sload']  * df['sjit']
    df['load_jit_interaction_dst'] = df['dload']  * df['djit']
    df['pkt_jit_interaction_src']  = df['spkts']  * df['sjit']
    df['pkt_jit_interaction_dst']  = df['dpkts']  * df['djit']

    # — 8) Statistical
    df['mean_pkt_size']    = (df['smeansz'] + df['dmeansz']) / 2
    df['tcp_seq_diff']     = df['stcpb'] - df['dtcpb']

    # — 9) Service & FTP command ratios
    df['ftp_cmd_flow_ratio']  = df['ct_ftp_cmd']      / (df['ct_flw_http_mthd'] + eps)
    df['service_src_ratio']   = df['ct_srv_src']      / (df['ct_state_ttl']     + eps)
    df['service_dst_ratio']   = df['ct_srv_dst']      / (df['ct_state_ttl']     + eps)

    # — 10) Response body rate
    df['res_bdy_rate']        = df['res_bdy_len']     / (df['duration']         + eps)

    # — 11) Time features (dari stime & ltime)
    df['stime_hour']        = df['stime'].dt.hour
    df['stime_weekday']     = df['stime'].dt.weekday
    df['ltime_hour']        = df['ltime'].dt.hour
    df['ltime_weekday']     = df['ltime'].dt.weekday

    return df


In [ ]:
train_df = generate_features(train_df)

In [ ]:
# ——— 8) DROP HIGHLY-CORRELATED FEATURES ———
# (a) Hitung korelasi antar numeric (termasuk FEA baru)
num_cols = train_df.select_dtypes(include=['int64','float64']).columns.tolist()
# Jangan sertakan target/label sendiri
for drop in ['attack_cat_enc','label']:
    if drop in num_cols: num_cols.remove(drop)

corr = train_df[num_cols].corr().abs()
upper = corr.where(np.triu(np.ones(corr.shape), k=1).astype(bool))

# threshold 0.90
to_drop_corr = [col for col in upper.columns if (upper[col] > 0.90).any()]
train_df.drop(columns=to_drop_corr, inplace=True)
print(f"Dropped {len(to_drop_corr)} highly-correlated features:", to_drop_corr)

Dropped 17 highly-correlated features: ['sloss', 'dloss', 'dpkts', 'dwin', 'synack', 'ackdat', 'ct_ftp_cmd', 'dst_pps', 'load_ratio', 'inter_pkt_ratio', 'tcp_setup_ratio', 'total_load', 'total_jitter', 'total_inter_pkt', 'total_tcp_setup', 'pkt_jit_interaction_dst', 'ftp_cmd_flow_ratio']


In [ ]:
# ——— 9) DROP NON-FEATURE COLUMNS ———
# IP / teks / timestamp asli bukan input numeric
non_feature_cols = [
    'srcip','dstip','service','state','proto',  # nominal asli
    'stime','ltime'                             # kalau sudah ekstrak jam/hari
]
present = [c for c in non_feature_cols if c in train_df]
train_df.drop(columns=present, inplace=True)
print("Dropped non-feature cols:", present)

Dropped non-feature cols: ['srcip', 'dstip', 'service', 'state', 'proto', 'stime', 'ltime']


In [ ]:
# ——— 10) ENCODE TARGET ———
from sklearn.preprocessing import LabelEncoder

# jika Anda ingin multi-class berdasarkan attack_cat
le = LabelEncoder()
train_df['attack_cat_enc'] = le.fit_transform(train_df['attack_cat'])
mapping = dict(zip(le.classes_, le.transform(le.classes_)))
print("attack_cat → attack_cat_enc mapping:", mapping)

# drop kolom string asli
train_df.drop(columns=['attack_cat'], inplace=True)

attack_cat → attack_cat_enc mapping: {'analysis': np.int64(0), 'backdoors': np.int64(1), 'dos exploits': np.int64(2), 'fuzzers': np.int64(3), 'generic': np.int64(4), 'normal': np.int64(5), 'reconnaissance': np.int64(6), 'shellcode': np.int64(7), 'worms': np.int64(8)}


In [ ]:
# ——— FINAL CHECK ———
print("Final shape:", train_df.shape)
print("Any missing values left?", train_df.isnull().any().any())
print("Dtypes breakdown:\n", train_df.dtypes.value_counts())

Final shape: (2057312, 61)
Any missing values left? False
Dtypes breakdown:
 int64      32
float64    25
int32       4
Name: count, dtype: int64


# EXPLORATORY DATA ANALYSIS (EDA)

In [ ]:
# plt.figure(figsize=(10, 5))
# sns.countplot(x='label', data=train_df)
# plt.title("Distribusi Label")
# plt.show()


### Interpretasi

Berdasarkan distribusi label dalam dataset, terlihat bahwa label 0 (normal) mendominasi sebesar 95.16%, sementara label 1 (serangan) hanya mencakup 4.84% dari seluruh data. Ketidakseimbangan (imbalance) ini menunjukkan bahwa dataset sangat didominasi oleh data normal dibanding data serangan. Hal ini berpotensi menjadi masalah serius saat melatih model machine learning, karena model bisa saja cenderung mengklasifikasikan hampir semua data sebagai normal (label 0) untuk mendapatkan akurasi tinggi secara semu, tanpa benar-benar belajar mendeteksi serangan.

Untuk mengatasi permasalahan ini, beberapa strategi perlu diterapkan. Pertama, gunakan metrik evaluasi yang sensitif terhadap ketidakseimbangan kelas seperti precision, recall, F1-score, dan ROC-AUC, bukan hanya mengandalkan akurasi. Kedua, terapkan teknik resampling seperti SMOTE (Synthetic Minority Over-sampling Technique) untuk menyeimbangkan jumlah sampel antar kelas, atau gunakan model yang mendukung class weighting, seperti RandomForestClassifier(class_weight='balanced') atau XGBoost dengan scale_pos_weight. Langkah-langkah ini penting agar model yang dibangun dapat mengenali pola serangan secara lebih efektif, meskipun datanya sangat minoritas.

In [ ]:
# plt.figure(figsize=(6,4))
# sns.countplot(x='attack_cat_enc', data=train_df, order=train_df['attack_cat_enc'].value_counts().index)
# plt.title("Target Distribution")
# plt.xticks(rotation=45)
# plt.tight_layout()
# plt.show()

In [ ]:
# sebelum Winsorizing/skew
num_cols = train_df.select_dtypes(include=['int64','float64']).columns.tolist()

# exclude target columns
for target in ['label','attack_cat_enc']:
    if target in num_cols:
        num_cols.remove(target)


In [ ]:
# # Distribution & skewness
# skew_info = {}
# for col in num_cols:
#     skew = train_df[col].skew()
#     skew_info[col] = skew
#     plt.figure(figsize=(5,2))
#     sns.histplot(train_df[col], bins=50, kde=True)
#     plt.title(f"{col} (skew={skew:.2f})")
#     plt.tight_layout()
#     plt.show()

In [ ]:
# # Outlier detection via boxplot
# for col in num_cols:
#     plt.figure(figsize=(5,2))
#     sns.boxplot(x=train_df[col])
#     plt.title(f"Boxplot {col}")
#     plt.tight_layout()
#     plt.show()

In [ ]:
# # ——— 4) CORRELATION MATRIX ———
# plt.figure(figsize=(12,10))
# corr = train_df[num_cols].corr()
# sns.heatmap(corr, cmap='vlag', center=0, fmt=".2f", square=True)
# plt.title("Correlation Matrix")
# plt.show()

In [ ]:
from sklearn.preprocessing import PowerTransformer, StandardScaler

# ——— A) Subset normal traffic ———
normal_df = train_df[train_df['label'] == 0].copy()
print("Normal only shape:", normal_df.shape)

Normal only shape: (1959668, 61)


In [ ]:
# ——— B) Tentukan kolom numerik (exclude target) ———
num_cols = normal_df.select_dtypes(include=[np.number]).columns.drop(
    ['label', 'attack_cat_enc'], errors='ignore'
).tolist()

In [ ]:
# ——— C) Winsorizing (median replacement) ———
for col in num_cols:
    q1, q3 = normal_df[col].quantile([0.25, 0.75])
    iqr = q3 - q1
    lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
    med = normal_df[col].median()
    normal_df[col] = np.where(
        normal_df[col] < lower, med,
        np.where(normal_df[col] > upper, med, normal_df[col])
    )

In [ ]:
# # ——— 6) ADVANCED OUTLIER HANDLING: WINSORIZING with MEDIAN REPLACEMENT ———
# # Definisi: ganti values di luar [Q1-1.5*IQR, Q3+1.5*IQR] menjadi median kolom
# for col in num_cols:
#     q1 = train_df[col].quantile(0.25)
#     q3 = train_df[col].quantile(0.75)
#     iqr = q3 - q1
#     lower, upper = q1 - 1.5*iqr, q3 + 1.5*iqr
#     median = train_df[col].median()
#     train_df[col] = np.where(train_df[col] < lower,  median,
#                      np.where(train_df[col] > upper,  median,
#                               train_df[col]))

In [ ]:
# # ——— 7) ATASI SKEWNESS: POWER TRANSFORMER (YEO-JOHNSON) ———
# # ini bisa mereduksi skew baik positif maupun negatif
# pt = PowerTransformer(method='yeo-johnson', standardize=True)
# train_df[num_cols] = pt.fit_transform(train_df[num_cols])

In [ ]:
# ——— D) Reduksi Skewness (Yeo–Johnson) ———
# a) Setup transformer & scaler
pt = PowerTransformer(method='yeo-johnson', standardize=False)
scaler = StandardScaler()

# b) Yeo–Johnson untuk semua fitur numerik (kecuali target)
normal_df[num_cols] = pt.fit_transform(normal_df[num_cols])

# c) Standardisasi (zero mean, unit variance)
normal_df[num_cols] = scaler.fit_transform(normal_df[num_cols])

In [ ]:
# d) Cek hasil transformasi awal
print("After YJ & standardize:")
print(" Skewness (abs > 0.5):")
print(normal_df[num_cols].skew().abs()
      .sort_values(ascending=False)
      .head(), "\n")
print(" Any NaNs? ", normal_df[num_cols].isna().any().any())
print(" Dtypes:", normal_df[num_cols].dtypes.unique(), "\n")

After YJ & standardize:
 Skewness (abs > 0.5):
Name
smeansz                     1.446125
swin                        0.934260
window_ratio                0.929448
dur                         0.914448
load_jit_interaction_src    0.812962
dtype: float64 

 Any NaNs?  False
 Dtypes: [dtype('float64')] 



In [ ]:
# ——— 7) Optional: Log‐Transform on Remaining Heavily‐Skewed Features ———
to_log = ['smeansz', 'dur']
print("Applying log1p to:", to_log)
for col in to_log:
    normal_df[col] = np.log1p(normal_df[col])

# Re‐scale only the log‐transformed cols
normal_df[to_log] = scaler.fit_transform(normal_df[to_log])

# Cek skewness baru
print("New skewness for log‐transformed cols:")
print(normal_df[to_log].skew(), "\n")

Applying log1p to: ['smeansz', 'dur']
New skewness for log‐transformed cols:
Name
smeansz   -0.606202
dur       -1.554149
dtype: float64 



In [ ]:
# ——— 8) Final Check ———
print("Final pipeline check:")
print(" Skewness (abs > 0.5):")
print(normal_df[num_cols].skew().abs()
      .sort_values(ascending=False)
      .head())
print(" Any NaNs? ", normal_df[num_cols].isna().any().any())

Final pipeline check:
 Skewness (abs > 0.5):
Name
dur                         1.554149
swin                        0.934260
window_ratio                0.929448
load_jit_interaction_src    0.812962
stcpb                       0.796010
dtype: float64
 Any NaNs?  True


In [ ]:
# Contoh: Yeo–Johnson saja untuk dur
from sklearn.preprocessing import PowerTransformer
pt = PowerTransformer(method='yeo-johnson', standardize=False)
normal_df['dur'] = pt.fit_transform(normal_df[['dur']])

# Lalu rescale
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
normal_df['dur'] = scaler.fit_transform(normal_df[['dur']])

# Cek skew
print("Durasi skew baru:", normal_df['dur'].skew())


Durasi skew baru: -0.043399784238973595


In [ ]:
# normal_df.to_csv('normal_df.csv', index=False, encoding='utf-8')

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# — 1) Definisikan X dan kedua target Y
#   • X: semua fitur kecuali kolom target
#   • y_attack: multiclass target (0…8)
#   • y_label : binary target (0 normal, 1 attack)
X = train_df.drop(columns=['attack_cat_enc', 'label'])
y_attack = train_df['attack_cat_enc']
y_label  = train_df['label']

# — 2) Split data sekaligus untuk kedua target
#    Kita stratify berdasarkan multiclass, agar semua kelas attack terwakili di train/test
X_train, X_test, \
y_train_attack, y_test_attack, \
y_train_label,  y_test_label = train_test_split(
    X, y_attack, y_label,
    test_size=0.2,
    stratify=y_attack,
    random_state=42
)

# — 3) Scaling fitur numeric
numeric_cols = X_train.select_dtypes(include=['int64','float64']).columns
scaler = StandardScaler()
X_train[numeric_cols] = scaler.fit_transform(X_train[numeric_cols])
X_test[numeric_cols]  = scaler.transform(X_test[numeric_cols])

# — 4) Cek bentuk data
print("X_train:", X_train.shape)
print("X_test :", X_test.shape)
print("y_train_attack distribution:\n", y_train_attack.value_counts(normalize=True))
print("y_train_label distribution:\n",  y_train_label.value_counts(normalize=True))


X_train: (1645849, 59)
X_test : (411463, 59)
y_train_attack distribution:
 attack_cat_enc
5    0.952538
2    0.015198
4    0.012335
3    0.010593
6    0.006493
0    0.001061
1    0.000964
7    0.000735
8    0.000083
Name: proportion, dtype: float64
y_train_label distribution:
 label
0    0.952538
1    0.047462
Name: proportion, dtype: float64


In [ ]:
from sklearn.metrics import classification_report

# Contoh multiclass
from sklearn.ensemble import RandomForestClassifier
clf_mc = RandomForestClassifier(random_state=42)
clf_mc.fit(X_train, y_train_attack)
print(classification_report(y_test_attack, clf_mc.predict(X_test)))

# Contoh binary
clf_bin = RandomForestClassifier(random_state=42)
clf_bin.fit(X_train, y_train_label)
y_pred_proba = clf_bin.predict_proba(X_test)[:,1]
y_pred = (y_pred_proba > 0.5).astype(int)
print(classification_report(y_test_label, y_pred))


KeyboardInterrupt: 

In [ ]:
from sklearn.model_selection import GridSearchCV

params = {
  'n_estimators': [100,200],
  'max_depth': [None, 10, 20],
  'min_samples_leaf': [1,5,10]
}
grid = GridSearchCV(RandomForestClassifier(random_state=42),
                    params, cv=3, scoring='f1_macro', n_jobs=-1)
grid.fit(X_train, y_train_attack)
print(grid.best_params_, grid.best_score_)


In [ ]:
from sklearn.metrics import ConfusionMatrixDisplay

ConfusionMatrixDisplay.from_estimator(
  clf_mc, X_test, y_test_attack, normalize='true', cmap='Blues'
)


In [ ]:
import pandas as pd
fi = pd.Series(clf_mc.feature_importances_, index=X_train.columns)
print(fi.nlargest(10))


In [ ]:
train_df.head()

In [ ]:
train_df['label'].value_counts()

In [ ]:
# ——— 2) PREPROCESSING SEDERHANA ———
# (Asumsikan kamu sudah punya fungsi preprocess() yang mencakup:
#  generate_features, fillna khusus, drop kolom non-fit, drop high-corr, encode attack_cat_enc & label, scaling, dsb.)
train_df = preprocess(train_df)
test_df  = preprocess(test_df)

# Pisahkan X / y
X_train = train_df.drop(columns=['attack_cat_enc','label'])
y_train_attack = train_df['attack_cat_enc']   # multiclass
y_train_label  = train_df['label']            # binary

X_test  = test_df.drop(columns=['attack_cat_enc','label'])
y_test_attack = test_df['attack_cat_enc']
y_test_label  = test_df['label']

# ——— 3) TRAIN MODEL ———
clf = RandomForestClassifier(
    n_estimators=200,
    max_depth=None,
    random_state=42,
    n_jobs=-1
)
clf.fit(X_train, y_train_attack)  # contoh untuk multiclass

# ——— 4) EVALUASI PADA TRAINING SET ———
y_pred_train_mc = clf.predict(X_train)
print("=== TRAINING SET (Multiclass) ===")
print("Accuracy :", accuracy_score(y_train_attack, y_pred_train_mc))
print(classification_report(y_train_attack, y_pred_train_mc, target_names=le.classes_))

# juga bisa ROC-AUC one-vs-rest jika mau:
# y_proba_train = clf.predict_proba(X_train)
# print("ROC-AUC (macro):", roc_auc_score(
#    y_train_attack, y_proba_train, average='macro', multi_class='ovr'
# ))

# ——— 5) EVALUASI PADA TESTING SET ———
y_pred_test_mc = clf.predict(X_test)
print("=== TESTING SET (Multiclass) ===")
print("Accuracy :", accuracy_score(y_test_attack, y_pred_test_mc))
print(classification_report(y_test_attack, y_pred_test_mc, target_names=le.classes_))

# ——— 6) (OPSIONAL) BINARY EVALUASI ———
# Jika kamu ingin lihat performa normal vs attack:
clf_bin = RandomForestClassifier(n_estimators=200, random_state=42, n_jobs=-1)
clf_bin.fit(X_train, y_train_label)

for split, X, y_true in [
    ("TRAIN", X_train, y_train_label),
    ("TEST",  X_test,  y_test_label)
]:
    y_pred = clf_bin.predict(X)
    y_proba = clf_bin.predict_proba(X)[:,1]
    print(f"=== {split} SET (Binary) ===")
    print("Accuracy :", accuracy_score(y_true, y_pred))
    print(classification_report(y_true, y_pred, target_names=['normal','attack']))
    print("ROC AUC  :", roc_auc_score(y_true, y_proba))
    print("Confusion Matrix:\n", confusion_matrix(y_true, y_pred))